In [1]:
import pandas as pd               #import library

In [3]:
#load dataset
df=pd.read_excel(r"C:\Users\Dell\Dataset1.csv.xlsx")
df.head()

,SN,Train_No,Station_Code,1A,2A,3A,SL,Station_Name,Route_Number,Arrival_time,Departure_Time,Distance
0,1,107,SWV,100,100,100,100,SAWANTWADI R,1,00:00:00,10:25:00,0
1,2,107,THVM,260,228,196,164,THIVIM,1,11:06:00,11:08:00,32
2,3,107,KRMI,345,296,247,198,KARMALI,1,11:28:00,11:30:00,49
3,4,107,MAO,490,412,334,256,MADGOAN JN.,1,12:10:00,00:00:00,78
4,1,108,MAO,100,100,100,100,MADGOAN JN.,1,00:00:00,20:30:00,0


In [4]:
#Task1.1:Overview of dataset-total records and attributes
df.shape

(186074, 12)

In [7]:
df.columns             #display column names

Index(['SN', 'Train_No', 'Station_Code', '1A', '2A', '3A', 'SL',
       'Station_Name', 'Route_Number', 'Arrival_time', 'Departure_Time',
       'Distance'],
      dtype='object')

In [8]:
#basic information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 186074 entries, 0 to 186073
Data columns (total 12 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   SN              186074 non-null  int64 
 1   Train_No        186074 non-null  int64 
 2   Station_Code    186074 non-null  object
 3   1A              186074 non-null  int64 
 4   2A              186074 non-null  int64 
 5   3A              186074 non-null  int64 
 6   SL              186074 non-null  int64 
 7   Station_Name    186074 non-null  object
 8   Route_Number    186074 non-null  int64 
 9   Arrival_time    186074 non-null  object
 10  Departure_Time  186074 non-null  object
 11  Distance        186074 non-null  int64 
dtypes: int64(8), object(4)
memory usage: 17.0+ MB


In [9]:
df.dtypes                      #data type of each column

SN                 int64
Train_No           int64
Station_Code      object
1A                 int64
2A                 int64
3A                 int64
SL                 int64
Station_Name      object
Route_Number       int64
Arrival_time      object
Departure_Time    object
Distance           int64
dtype: object

In [10]:
#Task1.2:List all trains with their starting and ending stations
df[['Train_No', 'Station_Code', 'Station_Name', 'Route_Number']].head(20)

,Train_No,Station_Code,Station_Name,Route_Number
0,107,SWV,SAWANTWADI R,1
1,107,THVM,THIVIM,1
2,107,KRMI,KARMALI,1
3,107,MAO,MADGOAN JN.,1
4,108,MAO,MADGOAN JN.,1
5,108,KRMI,KARMALI,1
6,108,THVM,THIVIM,1
7,108,SWV,SAWANTWADI R,1
8,128,MAO,MADGOAN JN.,1
9,128,KRMI,KARMALI,1


In [11]:
#Create a train route summary
train_routes = df.groupby('Train_No').agg(
    Starting_Station=('Station_Name', 'first'),
    Ending_Station=('Station_Name', 'last')).reset_index()
train_routes.head(10)

,Train_No,Starting_Station,Ending_Station
0,107,SAWANTWADI R,MADGOAN JN.
1,108,MADGOAN JN.,SAWANTWADI R
2,128,MADGOAN JN.,CHHATRAPATI
3,290,DELHI-SAFDAR,DELHI-SAFDAR
4,401,AURANGABAD,VARANASI JN.
5,421,LUCKNOW JN.,SHRI MATA VA
6,422,SHRI MATA VA,LUCKNOW JN.
7,477,SIRSA,SIRSA
8,502,RAJENDRANAGA,AMBALA CANTT
9,504,PATNA JN.,BATHINDA JN


In [12]:
#Include station codes too
train_routes = df.groupby('Train_No').agg(
    Starting_Station=('Station_Name', 'first'),
    Starting_Code=('Station_Code', 'first'),
    Ending_Station=('Station_Name', 'last'),
    Ending_Code=('Station_Code', 'last')).reset_index()
train_routes.head(10)

,Train_No,Starting_Station,Starting_Code,Ending_Station,Ending_Code
0,107,SAWANTWADI R,SWV,MADGOAN JN.,MAO
1,108,MADGOAN JN.,MAO,SAWANTWADI R,SWV
2,128,MADGOAN JN.,MAO,CHHATRAPATI,KOP
3,290,DELHI-SAFDAR,DSJ,DELHI-SAFDAR,DSJ
4,401,AURANGABAD,AWB,VARANASI JN.,BSB
5,421,LUCKNOW JN.,LKO,SHRI MATA VA,SVDK
6,422,SHRI MATA VA,SVDK,LUCKNOW JN.,LKO
7,477,SIRSA,SSA,SIRSA,SSA
8,502,RAJENDRANAGA,RJPB,AMBALA CANTT,UMB
9,504,PATNA JN.,PNBE,BATHINDA JN,BTI


In [13]:
#Check how many unique trains we have
print("Total unique trains:", df['Train_No'].nunique())

Total unique trains: 11113


In [27]:
#Task 1.3 — Calculate Number of Stops per Train
stops_per_train = df.groupby('Train_No').size().reset_index(
    name='Number_of_Stops')
stops_per_train

,Train_No,Number_of_Stops
0,107,4
1,108,4
2,128,22
3,290,14
4,401,12
...,...,...
11108,99904,12
11109,99905,11
11110,99906,12
11111,99907,12


In [24]:
# Top 10 trains with the most stops
stops_per_train.sort_values(by='Number_of_Stops',ascending=False).head(10)

,Train_No,Number_of_Stops
5977,53041,118
1348,13007,112
1372,13049,111
7383,58112,109
1349,13008,108
5978,53042,107
1373,13050,106
7382,58111,102
2431,19019,98
2436,19024,97


In [25]:
# Top 10 trains with the fewest stops
stops_per_train.sort_values(by='Number_of_Stops',ascending=True).head(10)

,Train_No,Number_of_Stops
10525,98330,2
10442,98160,2
10443,98161,2
10444,98162,2
10445,98163,2
10446,98164,2
10447,98165,2
10448,98166,2
10449,98167,2
10450,98168,2


In [31]:
#Task1.4: Identify trains with maximum and minimum stops

# maximum number of stops
max_stops = stops_per_train['Number_of_Stops'].max()
print("Maximum number of stops:", max_stops)

Maximum number of stops: 118


In [32]:
#Find the trains having maximum stops
max_stop_trains = stops_per_train[
    stops_per_train['Number_of_Stops'] == max_stops]
max_stop_trains

,Train_No,Number_of_Stops
5977,53041,118


In [33]:
#Find the minimum number of stops
min_stops = stops_per_train['Number_of_Stops'].min()
print("Minimum number of stops:", min_stops)

Minimum number of stops: 2


In [35]:
#Find the trains having minimum stops
min_stop_trains = stops_per_train[
    stops_per_train['Number_of_Stops'] == min_stops]
min_stop_trains

,Train_No,Number_of_Stops
11,557,2
53,3305,2
54,3306,2
105,5715,2
106,5716,2
...,...,...
11064,99499,2
11065,99500,2
11066,99501,2
11067,99502,2


## Level 1: Basic Data Review — Summary

The train schedule dataset contains **186,074 records and 12 attributes** with no missing values.

There are **11,113 unique trains** in the dataset. Each train has multiple station-level records, which were grouped by `Train_No` to calculate the number of station records/stops for each train.

The analysis shows that **Train 53041 has the maximum number of station records, with 118 stops**.

The minimum number of station records is **2**, and **1,249 trains** have this minimum value.

These findings provide a basic understanding of the train routes and station coverage in the dataset.